Update 2201.4.0 release note with OpenAPI changes (#6277)
ballerina-platform/ballerina-dev-website: downloads/swan-lake-release-notes/swan-lake-2201.4.0/RELEASE_NOTE.md

## Developer tools updates

### New features
#### OpenAPI Tool
Added support to generate Ballerina client and service declarations from Swagger 2.0(i.e. OpenAPI 2.0) definitions

#### Language Server

### Improvements

#### OpenAPI Tool

> - Added a mapping for the `additionalProperties` property in the OpenAPI object schema. With this support, the generated Ballerina records can be open or closed according to the `additionalProperties` details.

Reviewer: ```suggestion
- Added support for the `additionalProperties` attribute in OpenAPI object schemas. With this support, the generated Ballerina records can be either open or closed records, based on the `additionalProperties` details.
```

> >**Info:** Thereby, some of your existing generated Ballerina records may change into closed records after regenerating using the Ballerina 2201.4.0 (Swan Lake) version.

Reviewer: ```suggestion
>**Info:** Thereby, some of your already generated open records may change into closed records, when re-generating using 2201.4.0 (and above) versions.
```

> - Improved the OpenAPI schema `nullable:true` property support by providing generated a record field with `nil` as the default value (e.g. : `string name = ()` ), instead of making field both nilable and optional (e.g. : `string? name?;`).

Reviewer: ```suggestion
- Improved support for `nullable:true` property OpenAPI schema, to generate record fields with default values (e.g. `string? name = ();`), instead of making the field both nilable and optional (e.g. `string? name?;`).
```

> - Changed the request body and return types of the generated client remote/resource methods (from `json` to `http:Request` and `http:Response` types) for unhandled `Request`, and `Response` payload media types in the OpenAPI specification.

Reviewer: ```suggestion
- Changed the default request and response types of the generated Ballerina resource/remote methods from `json` to `http:Request` and `http:Response`, respectively.
```

## Developer tools updates

### New features
#### OpenAPI Tool
Added Swagger version 2 support for OpenAPI to Ballerina client, service generation.

#### Language Server

### Improvements

#### OpenAPI Tool

> - Added mapping for the `additionalProperty` property in the OpenAPI object schema. With this support, generated Ballerina records can be open records or closed records according to the `additionalProperty` property details.

Reviewer: @lnash94 What's the correct property name in here? is it `additionalProperty` or `additionalProperties`?

Author: yeah it should be a `additionalProperties` . will change

  Thereby, some of your existing generated Ballerina records may change into close records after regenerating using Ballerina 2201.4.0 (Swan Lake) version.

> - Improved OpenAPI schema `nullable:true` property support by providing generated record field with default of value nil, instead of making field both nilable and optional.

Reviewer 2: How about this explanation with examples since the explanation is bit confusing ? @lnash94 

```suggestion
- Improved OpenAPI schema `nullable:true` property support by providing generated record field with default of value nil (e.g. : `string name = ()` ), instead of making field both nilable and optional (e.g. : `string? name?;`).
```

Author: +1, and we can remove the additional Info section. https://github.com/ballerina-platform/ballerina-dev-website/pull/6277/files#r1088560147

## Developer tools updates

### New features
#### OpenAPI Tool

> Added Swagger version 2 support for OpenAPI to Ballerina client, service generation.

Reviewer 3: ```suggestion
Added Swagger version 2 support for the OpenAPI to Ballerina client and service generations.
```

Reviewer: @lnash94 @praneesha @aneeshafedo Shall we simply say `Added support to generate Ballerina client and service declarations from OpenAPI 2.0(i.e. swagger) definitions`? This will convey the idea more clearer IMO.

Author: How about using `Swagger 2.0` instead of `OpenAPI 2.0` since the term "Swagger 2.0" is more commonly used to refer to this version of the specification;
`Added support to generate Ballerina client and service declarations from Swagger 2.0(i.e. OpenAPI 2.0) definitions`

### Improvements

#### OpenAPI Tool

> - Added mapping for the `additionalProperty` property in the OpenAPI object schema. With this support, generated Ballerina records can be open records or closed records according to the `additionalProperty` property details.

Reviewer 3: ```suggestion
- Added a mapping for the `additionalProperty` property in the OpenAPI object schema. With this support, the generated Ballerina records can be open or closed according to the `additionalProperty` details.
```

>   Thereby, some of your existing generated Ballerina records may change into close records after regenerating using Ballerina 2201.4.0 (Swan Lake) version.

Reviewer 3: ```suggestion
  >**Info:** Thereby, some of your existing generated Ballerina records may change into closed records after regenerating using the Ballerina 2201.4.0 (Swan Lake) version.
```

> - Improved OpenAPI schema `nullable:true` property support by providing generated record field with default of value nil, instead of making field both nilable and optional.

Reviewer 3: ```suggestion
- Improved the OpenAPI schema `nullable:true` property support by providing a generated record field with `nil` as the default value instead of making the field both nilable and optional.
```
What is the correct word `nilable` or `nullable`?

>   Therefore, some of your existing generated Ballerina records field may change into default value nil using Ballerina 2201.4.0 (Swan Lake) version.

Reviewer 3: ```suggestion
  >**Info:** Therefore, the value of some of your existing generated Ballerina record fields may change into the default value (i.e., `nil`) when the Ballerina 2201.4.0 (Swan Lake) version is used.
```

Reviewer: @lnash94 Do we really need to mention the impact for this case, as there won't be any breaking cases (correct me if I'm wrong) for the end users?

Author: Yeah, It won't be any breaking change. I added this in case the user gets confused with generated code after seeing the default nil assigning.
@aneeshafedo appreciate any thoughts regarding the connector behaviour ?

Reviewer 2: Yes it won't be a breaking change for a connector user. Only when some connector is regenerated using the 2201.4.0 he will notice the changes in those records.

> - Changed generated client remote/resource methods' request body and return types, `json` into `http:Request` and `http:Response` types for unhandled Request, Response payload media types in OpenAPI specification.

Reviewer 3: ```suggestion
- Changed the request body and return types of the generated client remote/resource methods (from `json` to `http:Request` and `http:Response` types) for unhandled `Request`, and `Response` payload media types in the OpenAPI specification.
```
Check if the technical meaning is affected by the suggestions.

Reviewer: @lnash94 Is this done only for the service generation side?

Author: This is done for client generation only. Service generation is yet to be done.
